### ISSUE TITLE
Problems with (cleaning up) shared memory on large supercomputers

### ISSUE THREAD

**Issue by earnric** [pynbody/pynbody#356]
Not sure if this is a pynbody issue, but I have been trying to load some large ramses snaps (output dirs) for a big run... I'm just loading the snap and trying to write out some properties (like axep, boxsize, etc.). However, the python is crashing and (I'm being told) is leaving very large "pynbody*" files in 

/dev/shm

The shared memory partition. This is causing problem and the need for the administrators to manually clean up /dev/shm after.

I'm running the pynbody python program on system with 64 128 GB/node -- and I'm requesting 5 nodes so I have a ton of RAM -- but it still fails (about 75% of the time -- the other 25% it finishes).

I guess this issue is a suggestion to look at releasing data structures, etc. whenever they are no longer needed. I'm going to need to do more pynbody processing on these large snaps and I don't think I can make it work: even with more than 500 GB of RAM at my disposal...

**Comment by apontzen** [comment]
You could uninstall the posix_ipc module if it's important to you not to be left with dangling shared memory segments, or you could just delete the leftover shared memory yourself (you should not need an administrator to do this for you).

Otherwise unfortunately there's not much that can be done about this other than find the issue that's causing your segfault.

A bit of background - the need for shared memory comes from the Python GIL forcing the ramses parallel read into multiple processes, rather than threads. Unfortunately if CPython crashes, there is no mechanism to release this shared memory. Pynbody implements an atexit handler but this will only catch memory leaks for "benign" exits: see http://stackoverflow.com/questions/40013946/how-to-clean-up-upon-a-crash

The third alternative would be to reimplement the ramses reader in C, which would allow the parallel read to be executed on threads rather than in processes. However this is a significant task.

**Comment by earnric** [comment]
Hum... thx Andrew. I'm tempted to build without posix, but with my large snaps (~450 GB !!) I fear it would take forever to load.(??)

Thx.

**Comment by apontzen** [comment]
Yes, that would be slow I would imagine. I do think your best option is to track down what causes the segfault - obviously that will be a bug either in pynbody or elsewhere. Use something like gdb to get a stack backtrace.

In the meantime, just delete those /dev/shm files. You could even write a wrap-around script to do it automatically, rm /dev/shm/pynbody* at the end of each run.

**Comment by earnric** [comment]
Yea… that’s what I’ve done (rm /dev…) …

Thx.


On Feb 3, 2017, at 11:06 AM, Andrew Pontzen <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, that would be slow I would imagine. I do think your best option is to track down what causes the segfault - obviously that will be a bug either in pynbody or elsewhere. Use something like gdb to get a stack backtrace.

In the meantime, just delete those /dev/shm files. You could even write a wrap-around script to do it automatically, rm /dev/shm/pynbody* at the end of each run.

**Comment by biernackip** [comment]
I'm not sure this would help, but maybe you can check RAMSES Read++ by Oliver Hahn: https://bitbucket.org/ohahn/ramsesread

This maybe would help to rewrite RAMSES reader...

**Comment by earnric** [comment]
Thanks! I may be able to use that directly to get the info out...

**Comment by earnric** [comment]
I can't seem to clone the Ramses reader project? Neither HTTPS and SSH seems to work...

**Comment by kyleaoman** [comment]
Ah that's unfortunate, oh well.
